gk/main.c

```diff
@@ -865,21 +978,40 @@ process_pkts_front(uint8_t port_front, uint8_t port_back,
 			switch (fib->action) {
 			case GK_FWD_GRANTOR:
 				/*
+				 * We randomly drop entries to
+				 * alleviate memory pressure
+				 * when the table is full.
+				 *
 				 * The entry instructs to enforce
 				 * policies over its packets,
 			 	 * initialize an entry in the
 				 * flow table, proceed as the
 				 * brand-new entry instructs, and
 			 	 * go to the next packet.
 			 	 */
+hash_add_pkt_flow:
 				ret = rte_hash_add_key_with_hash(
 					instance->ip_flow_hash_table,
  					&packet.flow, pkt->hash.rss);
-				if (ret < 0) {
+				if (ret == -EINVAL) {
 					RTE_LOG(ERR, HASH,
-						"The GK block failed to add new key to hash table!\n");
+						"The GK block failed to add new key to hash table in %s with invalid parameters!\n",
+						__func__);
 					drop_packet(pkt);
 					continue;
+				} else if (ret == -ENOSPC) {
+					RTE_LOG(WARNING, HASH,
+						"The GK block failed to add new key to hash table in %s due to lack of space!\n",
+						__func__);
+					ret = drop_flow_entry_randomly(
+						instance,
+						gk_conf->flow_ht_size);
+					if (ret < 0) {
+						drop_packet(pkt);
+						continue;
+					}
+
+					goto hash_add_pkt_flow;
```

Comment:
In case we need to drop the incoming request flow, we should call `gk_process_request()` on it before freeing its memory to give it a chance to reach a Grantor server. Once the code is here to do this, add a comment explaining it since the why won't be obvious.

Reply:
Since ``gk_process_request()`` requires to specify the parameter ``struct flow_entry *fe``, however, at this step, we cannot find the corresponding flow entry for this packet, should we modify ``gk_process_request()`` to allow ``NULL`` parameter for the flow entry?

Reply:
I'd avoid changing `gk_process_request()` as you suggested because it adds one more special case to the code, that is, it increases complexity. A solution that doesn't add complexity is to have a local variable `struct flow_entry temp_fe`, initialize it, and pass it to `gk_process_request()`.